In [1]:
# dictionary
myDict = {'≤':'\leq','⊕':'\oplus','⊘':'\oslash ','σ':'\sigma', \
    '≠':'\neq', '⊖':'\ominus', '⊗':'\otimes', 'δ':'\delta', 'ϵ': '\epsilon', '≈':'\approx', \
        '𝐱':'\mathbf{x}', '𝐯':'\mathbf{v}', '𝐞':'\mathbf{e}','𝐲':'\mathbf{y}', '∈':'\in', 'ℝ':'\mathbb{R}'}

In [3]:
# read in the file
# replacing the unicode characters (actuallly not necessary)
filename = "newPS.rmarkdown"
with  open(filename, encoding='utf-8') as infile, open('newPS2.rmarkdown', 'w', encoding='utf-8') as outfile:
    for line in infile:
        for old, new in myDict.items():
            line = line.replace(old, new)
        outfile.write(line)

In [52]:
# A function that replaces matches of string with appropriate colorings

fpDict = {16:[15,5,10], 32:[127,8,23], 64:[1023,11,52]}

def colorBits(match):
    '''Returns a copy of colored string from rergex match.
    '''
    str = match.group(0)
    arr = list(str)
    l = len(arr) # determine bit length
    Q = fpDict[l-2][1]
    S = fpDict[l-2][2]

    sign_bit = [r'$\color{red}{', arr[1], r'}'] # string with escape character
    expo_bit = [r'\color{green}{'] + arr[1:Q+1] + [r'}']
    mantissa = [r'\color{blue}{'] + arr[Q+1:l-1] + [r'}$']

    return ''.join(sign_bit+expo_bit+mantissa)

In [21]:
colorBits("$01110010101110000000000000000000$")

$\color{red}{0}\color{green}{01110010}\color{blue}{101110000000000000000000}$


Just to test $\color{red}{0}\color{green}{01110010}\color{blue}{101110000000000000000000}$

In [53]:
# regular expression stuff
import re
text = "an $01000000101110000000000000000000$  $1000001$"

pattern = re.compile(r'([$])+([0,1]{16,})([$])') #ensures at least 32-bit

matches  = pattern.finditer(text)
for match in matches:
    print(match.group(0))
    print(colorBits(match))


$01000000101110000000000000000000$
$\color{red}{0}\color{green}{01000000}\color{blue}{101110000000000000000000}$


In [69]:
# Use re.sub function
original = "and $132 = (10000100)_2$ hence: $01000010000000000000000000000000$  "
replaced = re.sub(r'([$])+([0,1]{16,})([$])', colorBits, original)
print(replaced)

and $132 = (10000100)_2$ hence: $\color{red}{0}\color{green}{01000010}\color{blue}{000000000000000000000000}$  


In [71]:
with open('newPS2.rmarkdown', 'r', encoding="utf-8") as f:
    contents = f.read()
    replaced = re.sub(r'([$])+([0,1]{16,})([$])', colorBits, contents)

with open('PSCollection.markdown', 'w') as g:
    g.write(replaced)